File: Java/src/Trabalhador.java
public class Trabalhador extends Thread{
    private Arquivo arquivoCompartilhado;
    private Tarefa tarefa;
    private Executor executor;
    
    private long tempoInicio;

    public Trabalhador(Executor executor, Arquivo arquivoCompartilhado){
        this.executor = executor;
        this.arquivoCompartilhado = arquivoCompartilhado;
    }
    
    public void escrita(){
        try {
            sleep((long)tarefa.getCusto() * 1000);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }

        int valorResultado = arquivoCompartilhado.escrita(tarefa.getValor());
        executor.guardarResultado(new Resultado(tarefa.getIdTarefa(), valorResultado, (System.currentTimeMillis() - tempoInicio) ));
    }

    public void leitura(){
        try {
            sleep((long)tarefa.getCusto() * 1000);
        } catch (InterruptedException e) {
            e.printStackTrace();  
        }
        
        int valorResultado = arquivoCompartilhado.leitura();
        executor.guardarResultado(new Resultado(tarefa.getIdTarefa(), valorResultado, (System.currentTimeMillis() - tempoInicio) ));
    }

    public void run(){
        tarefa = executor.pegarTarefa();
        while(tarefa != null){
            tempoInicio = System.currentTimeMillis();

            if (tarefa.getTipo() == 0) {
                escrita();
            }
            else {
                leitura();
            }

            tarefa = executor.pegarTarefa();
        }
    }
}

File: GO/src/classes/Trabalhador.go
package classes

import "time"

type Trabalhador struct {
	ArquivoCompartilhado *Arquivo
	Tarefa               *Tarefa
	Executor             *Executor
	TempoInicio          time.Time
	CanalBufferizado     <-chan *Tarefa
}

func CriarTrabalhador(arquivoCompartilhado *Arquivo, executor *Executor, canalBufferizado <-chan *Tarefa) *Trabalhador{
	return &Trabalhador{
		ArquivoCompartilhado: arquivoCompartilhado,
		Executor: executor,
		CanalBufferizado: canalBufferizado,
	}
}

func (t *Trabalhador) Run() {
	for tarefa := range t.CanalBufferizado {
		t.Tarefa = tarefa
		t.TempoInicio = time.Now()

		if t.Tarefa.Tipo == 0 {
			t.Escrita()
		} else {
			t.Leitura()
		}

	}
}

func (t *Trabalhador) Escrita() {
	//time.Sleep(time.Duration(t.Tarefa.Custo * float64(time.Second))/1000)

	valorResultado := t.ArquivoCompartilhado.Escrita(t.Tarefa.Valor)
	resultado := CriarResultado(t.Tarefa.IdTarefa, valorResultado, time.Since(t.TempoInicio).Milliseconds())

	t.Executor.GuardarResultado(resultado)
}

func (t *Trabalhador) Leitura() {
	//time.Sleep(time.Duration(t.Tarefa.Custo * float64(time.Second))/1000)

	valorResultado := t.ArquivoCompartilhado.Leitura()
	resultado := CriarResultado(t.Tarefa.IdTarefa, valorResultado, time.Since(t.TempoInicio).Milliseconds())

	t.Executor.GuardarResultado(resultado)
}
